src/odemis/util/comp.py

```diff
@@ -162,22 +162,16 @@ def generate_zlevels(focuser: "Actuator", zrange: Tuple[float, float], zstep: fl
     if n > MAX_ZLEVELS:
         raise IndexError(f"The number of zlevels, {n}, is too large. Reduce the zstep value to < {MAX_ZLEVELS}.")
 
-    # zstep often doesn't exactly fit to have a round number of zlevels. So try
-    # the floor and ceil values for the number of zlevels, and take the one that
-    # give a step closest to the requested zstep.
+    # When the zstep is bigger than the z range
+    # we return the z range itself
     n_f = math.floor(n)
-    n_c = math.ceil(n)
-
-    if n_f <= 1:  # would cause a division by zero
-        # => do the minimum amount of steps
-        n = 2
-    else:
-        # errors
-        ef = abs((zrange_abs[1] - zrange_abs[0]) / (n_f - 1) - abs(zstep))
-        ec = abs((zrange_abs[1] - zrange_abs[0]) / (n_c - 1) - abs(zstep))
-        n = n_f if ef < ec else n_c
-
-    if zstep > 0:
-        return numpy.linspace(zrange_abs[0], zrange_abs[1], n).tolist()
-    elif zstep < 0:
-        return numpy.linspace(zrange_abs[1], zrange_abs[0], n).tolist()
+    if n_f <= 1:
+        return [zrange_abs[0], zrange_abs[1]]
+
+    abs_zstep = abs(zstep)
+    focus_array_min = numpy.arange(focuser_pos, zrange_abs[0] - abs_zstep/10, -abs_zstep)
```

Comment:
Note that there is a warning about using arange with floats in the numpy documentation, it's recommended to use linspace instead. Linspace also allows for a `endpoint=True` argument, making it unnecessary to do the `-abs_zstep/10`
https://numpy.org/doc/stable/reference/generated/numpy.arange.html

![image](https://github.com/user-attachments/assets/06d4398b-320f-4a9f-ad1a-d1f02d6fecd0)

Reply:
I would suggest to write it like (note that the +1 for the number of steps is because you're including the endpoint):
```suggestion
    focus_array_min = numpy.linspace(zrange_abs[0], focuser_pos, (zrange_abs[0] + focuser_pos) / abs_zstep)
    focus_array_max = numpy.linspace(focuser_pos, zrange_abs[1], (zrange_abs[1] + focuser_pos) / abs_zstep + 1, endpoint=True)
```

Reply:
Note that arange and linspace don't give the same results. The previous implementation used linspace, but the problem is that it didn't respect the zstep. See such example:
```
$ numpy.arange(0, 10., 3)
 array([0., 3., 6., 9.])

$ numpy.linspace(0, 10, 4)
array([ 0.        ,  3.33333333,  6.66666667, 10.        ])
```
As I understand it, Karishma did handle the "floating point" issue by adding 10% of zstep . Of course, it should also be possible to write with linspace, but it'd be less explicit. I'd say the only thing missing here is to explain the trick of using `zstep/10`.